controllers/cloudstackcluster_controller.go

```diff
@@ -18,149 +18,141 @@ package controllers
+func NewCSClusterReconciliationRunner() *CloudStackClusterReconciliationRunner {
+	// Set concrete type and init pointers.
+	r := &CloudStackClusterReconciliationRunner{ReconciliationSubject: &infrav1.CloudStackCluster{}}
+	r.Zones = &infrav1.CloudStackZoneList{}
+	// Setup the base runner. Initializes pointers and links reconciliation methods.
+	r.ReconciliationRunner = csCtrlrUtils.NewRunner(r, r.ReconciliationSubject)
+	// For the CloudStackCluster, the ReconciliationSubject is the CSCluster
+	// Have to do after or the setup method will overwrite the link.
+	r.CSCluster = r.ReconciliationSubject
+
+	return r
 }
 
-// Actually reconcile cluster.
-func (r *CloudStackClusterReconciler) reconcile(
-	ctx context.Context,
-	log logr.Logger,
-	csCluster *infrav1.CloudStackCluster,
-) (ctrl.Result, error) {
+// Reconcile is the method k8s will call upon a reconciliation request.
+func (reconciler *CloudStackClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (retRes ctrl.Result, retErr error) {
+	return NewCSClusterReconciliationRunner().
+		UsingBaseReconciler(reconciler.ReconcilerBase).
+		ForRequest(req).
+		WithRequestCtx(ctx).
+		RunBaseReconciliationStages()
+}
 
-	log.V(1).Info("Reconciling CloudStackCluster.", "clusterSpec", csCluster.Spec)
+// Reconcile actually reconciles the CloudStackCluster.
+func (r *CloudStackClusterReconciliationRunner) Reconcile() (res ctrl.Result, reterr error) {
+	return r.RunReconciliationStages(
+		r.RequeueIfMissingBaseCRDs,
+		r.CreateZones(r.CSCluster.Spec.Zones),
+		r.CheckOwnedCRDsForReadiness(infrav1.GroupVersion.WithKind("CloudStackZone")),
+		r.GetZones(r.Zones),
+		r.VerifyZoneCRDs,
+		r.SetFailureDomains,
+		r.ResolveClusterDetails)
+}
 
-	// Prevent premature deletion of the csCluster construct from CAPI.
-	controllerutil.AddFinalizer(csCluster, infrav1.ClusterFinalizer)
-	// Set ready status so that a partial reconcile can be patched.
-	// Ready is required, and patching will fail otherwise.
-	csCluster.Status.Ready = false
+// ResolveClusterDetails fetches cluster specific details like domain and account IDs.
+func (r *CloudStackClusterReconciliationRunner) ResolveClusterDetails() (ctrl.Result, error) {
+	// Ensure that CAPI won't prematurely delete this CloudStackCluster.
+	controllerutil.AddFinalizer(r.CSCluster, infrav1.ClusterFinalizer)
 
 	// Create and or fetch cluster components.
-	err := r.CS.GetOrCreateCluster(csCluster)
+	err := r.CS.GetOrCreateCluster(r.CSCluster)
 	if err == nil {
-		log.Info("Fetched cluster info successfully.")
-		log.V(1).Info("Post fetch cluster status.", "clusterStatus", csCluster.Status)
+		r.Log.Info("Fetched cluster info successfully.")
+		r.Log.V(1).Info("Post fetch cluster status.", "clusterStatus", r.ReconciliationSubject.Status)
 
 		// Set cluster to ready to indicate readiness to CAPI.
-		csCluster.Status.Ready = true
+		if len(r.ReconciliationSubject.Status.FailureDomains) == 0 {
+			return r.RequeueWithMessage("blah")
```

Comment:
Please change message from blah to blcch.  (Or maybe something more elegant).

Reply:
I'll kill two birds with one stone and just delete the extraneous check. It was more of a debug helper than a permanently needed check.